### PR TITLE
Proof of concept to error when internal meta props are accessed directly

### PR DESCRIPTION
**Pull request by bor0**

### CODE REVIEW

**Comment by mikejolley** [comment]
@bor0 What if we just mapped it for the user with a notice rather than hard error?

**Comment by bor0** [comment]
@mikejolley, you mean calling `wc_doing_it_wrong`? I guess we can do that, too

**Comment by mikejolley** [comment]
@bor0 Yeah but also mapping the meta so it actually sets.

**Comment by bor0** [comment]
@mikejolley updated PR, one more round please.

File: includes/abstracts/abstract-wc-data.php [hunk 1]
```diff
@@ -269,6 +269,22 @@ public function get_meta_data() {
...
 	/**
+	 * Validate meta key.
+	 *
+	 * @since  3.x
+	 * @param  string $key
+	 * @return bool   true if validation is successful, false otherwise
+	 */
+	private function validate_meta_key( $key ) {
+		if ( $this->data_store && ! empty( $key ) && in_array( $key, $this->data_store->get_internal_meta_keys() ) ) {
+			wc_doing_it_wrong( __FUNCTION__, __( 'Meta properties should not be accessed directly. Use getters and setters.', 'woocommerce' ), '3.x' );
```
Comment (bor0) [inline includes/abstracts/abstract-wc-data.php]: It's `3.x` here. Need to change depending on which version this goes in

File: includes/abstracts/abstract-wc-data.php [hunk 2]
```diff
@@ -269,6 +269,22 @@ public function get_meta_data() {
 	}
 
 	/**
+	 * Validate meta key.
+	 *
+	 * @since  3.2.0
+	 * @param  string $key
+	 * @return bool   true if validation is successful, false otherwise
+	 */
+	private function validate_meta_key( $key ) {
```
Comment (mikejolley) [inline includes/abstracts/abstract-wc-data.php]: This should be protected.

File: includes/abstracts/abstract-wc-data.php [hunk 3]
```diff
@@ -278,6 +294,14 @@ public function get_meta_data() {
 	 * @return mixed
 	 */
 	public function get_meta( $key = '', $single = true, $context = 'view' ) {
+		if ( ! $this->validate_meta_key( $key ) ) {
```
Comment (mikejolley) [inline includes/abstracts/abstract-wc-data.php]: Given the context, would this method be better named `is_internal_meta_key` or similar?

File: includes/abstracts/abstract-wc-data.php [hunk 4]
```diff
@@ -278,6 +294,14 @@ public function get_meta_data() {
...
+			$function = 'get_' . $key;
```
Comment (mikejolley) [inline includes/abstracts/abstract-wc-data.php]: Some internal meta keys do not match the new prop names - what then?

Comment (bor0) [inline includes/abstracts/abstract-wc-data.php]: This is our best guess, trying to help them. If they are using `get_meta` where they shouldn't and also their getter is named weirdly, I don't think we can do much

**Pull request merged**